analysis/src/vm/romgen.rs

```diff
@@ -0,0 +1,217 @@
+//! Generate one ROM per machine from all declared functions
+
+use ast::asm_analysis::{
+    AnalysisASMFile, Batch, Incompatible, IncompatibleSet, Machine, PilBlock, Rom,
+};
+use number::FieldElement;
+
+use crate::utils::{
+    parse_function_statement, parse_instruction_body, parse_instruction_definition,
+    parse_pil_statement,
+};
+
+/// Generate the ROM for each machine based on its functions
+pub fn generate_rom<T: FieldElement>(file: AnalysisASMFile<T>) -> AnalysisASMFile<T> {
+    AnalysisASMFile {
+        machines: file
+            .machines
+            .into_iter()
+            .map(|(name, m)| (name, generate_machine_rom(m)))
+            .collect(),
+    }
+}
+
+fn generate_machine_rom<T: FieldElement>(mut machine: Machine<T>) -> Machine<T> {
+    if !machine.has_pc() {
+        // do nothing, there is not rom to be generated
+        machine
+    } else {
+        let function_id = "_function_id";
+
+        let pc = machine.pc().unwrap();
+
+        // add the necessary embedded instructions
+        let embedded_instructions = [
+            parse_instruction_definition(&format!(
+                "instr _jump_to_operation {{ {pc}' = {function_id} }}",
+            )),
+            parse_instruction_definition(&format!(
+                "instr _reset {{ {} }}",
+                machine
+                    .write_registers()
+                    .map(|r| format!("{}' = 0", r.name))
+                    .collect::<Vec<_>>()
+                    .join(", ")
+            )),
+            parse_instruction_definition(&format!("instr _loop {{ {}' = {} }}", pc, pc)),
+        ];
+
+        machine.instructions.extend(embedded_instructions);
+
+        // implement the return instruction. For this dispatcher, returning jumps to 0
+        machine
+            .instructions
+            .iter_mut()
+            .find(|i| i.name == "return")
+            .unwrap()
+            .body = parse_instruction_body(&format!("{{ {}' = 0 }}", machine.pc().unwrap()));
+
+        // generate the rom
+        // the functions are already batched, we just batch the dispatcher manually here
+        // we could also run batching again on the dispatcher
+
+        let mut rom: Vec<Batch<T>> = vec![];
+
+        // add the beginning of the dispatcher
+        rom.extend(vec![
+            Batch::from(vec![
+                parse_function_statement("_start::"),
+                parse_function_statement("_reset;"),
+            ])
+            .reason(IncompatibleSet::from(Incompatible::Unimplemented)),
+            // the latch is constrained to start at 0 using _L1
+            Batch::from(vec![parse_function_statement("_jump_to_operation;")])
+                .reason(IncompatibleSet::from(Incompatible::Label)),
+        ]);
+
+        // add each function, setting the function_id to the current position in the ROM
```

Comment:
where is this happening? I don't see `function_id` being used in the loop

Reply:
```rust
function.id = Some(T::from(rom.len() as u64));
```